proxy/src/main.rs

```diff
@@ -0,0 +1,163 @@
+use anyhow::{anyhow, bail, Result};
+use reqwest::blocking::{Client, Response};
+use reqwest::header::{self, HeaderMap};
+use reqwest::Method;
+use serde::{Deserialize, Serialize};
+use std::collections::HashMap;
+use std::process::ExitCode;
+use std::str::FromStr;
+use std::time::Duration;
+use std::{env, io};
+use url::Url;
+
+const ENV_CONFIG: &str = "SD_PROXY_ORIGIN";
+
+/// Incoming requests (as JSON) received over stdin
+#[derive(Deserialize, Debug)]
+#[serde(deny_unknown_fields)]
+struct IncomingRequest {
+    method: String,
+    path_query: String,
+    stream: bool,
+    #[serde(default)]
+    headers: HashMap<String, String>,
+    body: Option<String>,
+    #[serde(default = "default_timeout")]
+    timeout: u64,
+}
+
+/// Default timeout for requests; serde requires this be a function
+fn default_timeout() -> u64 {
+    10
+}
+
+/// Serialization format for non-streamed requests
+#[derive(Serialize, Debug)]
+struct OutgoingResponse {
+    status: u16,
+    headers: HashMap<String, String>,
+    body: String,
+}
+
+/// In a streamed response, we emit the checksum after the streaming finishes
+#[derive(Serialize, Debug)]
+struct StreamMetadataResponse {
+    sha256sum: String,
+    filename: String,
+}
+
+/// Serialization format for errors, always over stderr
+#[derive(Serialize, Debug)]
+struct ErrorResponse {
+    error: String,
+}
+
+fn handle_json_response(resp: Response) -> Result<()> {
+    let mut headers = HashMap::new();
+    for (name, value) in resp.headers() {
+        headers.insert(name.to_string(), value.to_str()?.to_string());
+    }
+    let outgoing_response = OutgoingResponse {
+        status: resp.status().as_u16(),
+        headers,
+        body: resp.text()?,
+    };
+    println!("{}", serde_json::to_string(&outgoing_response)?);
+    Ok(())
+}
+
+fn handle_stream_response(mut resp: Response) -> Result<()> {
+    // Get the headers, will be output later but we want to fail early if it's missing/invalid
+    let etag = resp
+        .headers()
+        .get(header::ETAG)
+        .ok_or_else(|| anyhow!("missing etag header"))?
+        .to_str()?
+        .to_string();
+    let content_disposition = resp
+        .headers()
+        .get(header::CONTENT_DISPOSITION)
+        .ok_or_else(|| anyhow!("missing content-disposition"))?
+        .to_str()?
+        .to_string();
+    let mut stdout = io::stdout().lock();
+    resp.copy_to(&mut stdout)?;
+    // Emit the checksum as stderr
+    eprintln!(
+        "{}",
+        serde_json::to_string(&StreamMetadataResponse {
+            sha256sum: etag,
```

Comment:
By the HTTP/1.1 spec ([RFC 7232 §2.3](https://datatracker.ietf.org/doc/html/rfc7232#section-2.3)), the `ETag` is "an opaque validator for differentiating between multiple representations of the same resource".  That's it.  It looks like Amazon S3 has [popularized](https://docs.aws.amazon.com/AmazonS3/latest/userguide/checking-object-integrity.html) the use of pairing `PUT` with `Content-MD5` → `GET` with matching `ETag`.  And we send the `submission.checksum` as the `ETag` since freedomofpress/securedrop#4314.  That's fair enough; a hash digest certainly lets you "differentiate between multiple representations".  :-)  But then it's not an opaque string; it has a structure like [`$hash_name:$hex_encoded_hash_value`](https://github.com/freedomofpress/securedrop/blob/8010052d516d3cacc4032ac7e05f32099aa38cd3/securedrop/models.py#L189).

I think we have three options here:

1. **Keep the proxy origin-agnostic:** pass `ETag` transparently, without interpreting it as any particular kind of hash, and let the downstream (i.e., the SDK/Client) handle that.
2. **Loosely couple the proxy to the Journalist API:** interpret the `ETag` as a hash if it matches this structure, or have the Journalist API set another header that can be interpreted this way in addition to an opaque `ETag`.
3. **Tightly couple the proxy to the Journalist API:** interpret the `ETag` as a hash if it matches this structure and fail otherwise.

I favor (1); I don't think the proxy should have these expectations for what it's talking to or for the format of the headers it's proxying.  But this points to your bigger question (in Slack) of how we should handle streaming and headers in this architecture.  I'll get to that now.

Reply:
This speaks directly to our situation:

https://stackoverflow.com/questions/58102135/is-there-a-standard-for-http-non-opaque-hash-based-etag-headers